sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,519 @@
+#pragma warning restore // ReadAsync(Memory<>) overload is not available in all targets
+
+            var countToLog = result;
+            DecrementLength(ref countToLog);
+            LogBuffer(buffer, offset, countToLog);
+
+            return result;
+        }
+
+        public override bool CanRead => _originalStream.CanRead;
+        public override bool CanSeek => _originalStream.CanSeek;
+        public override long Length => _originalStream.Length;
+        public override long Position
+        {
+            get => _originalStream.Position;
+            set => _originalStream.Position = value;
+        }
+
+        public override void Close()
+        {
+            _originalStream.Close();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+
+            _originalStream.Dispose();
+        }
+
+        private void DecrementLength(ref int count)
+        {
+            var left = Math.Min(count, _maxLoggedBytes);
+            count = left;
+            _maxLoggedBytes -= count;
+        }
+
+        public override bool CanWrite => false;
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Flush()
+        {
+            // Flush is allowed on read-only stream
+        }
+    }
+    #endregion
+
+    #region ContentEventSourceWrapper
+
+    private readonly struct ContentEventSourceWrapper
+    {
+        private const int CopyBufferSize = 8 * 1024;
+        private readonly ClientModelEventSource? _eventSource;
+
+        private readonly int _maxLength;
+
+        private readonly CancellationToken _cancellationToken;
+
+        public ContentEventSourceWrapper(ClientModelEventSource eventSource, bool logContent, int maxLength, CancellationToken cancellationToken)
+        {
+            _eventSource = logContent ? eventSource : null;
+            _maxLength = maxLength;
+            _cancellationToken = cancellationToken;
+        }
+
+        public async ValueTask LogAsync(string requestId, bool isError, Stream? stream, Encoding? textEncoding, bool async)
+        {
+            EventType eventType = ResponseOrError(isError);
+
+            if (stream == null || !IsEnabled(eventType))
+            {
+                return;
+            }
+
+            var bytes = await FormatAsync(stream, async).ConfigureAwait(false).EnsureCompleted(async);
+            Log(requestId, eventType, bytes, textEncoding);
+        }
+
+        public async ValueTask LogAsync(string requestId, BinaryContent? content, Encoding? textEncoding, bool async)
+        {
+            EventType eventType = EventType.Request;
```

Comment:
Curious: why don't we call ResponseOrError in the BinaryContent overload? (where we do in the Stream overload, i.e. on line 346 above)